### PR TITLE
Third party revocations support

### PR DESCRIPTION
**Pull request by zapu**
Work in progress

This PR shall enable us to parse third party signatures, but we are mainly interested with designated revocations. To do this properly, we also parse `T.direcy`, which key owners use to mark 3rd party key ids that can revoke keys.

it breaks a bunch of tests right now, and the `skip_verify` things is not great. We don't want to verify signatures that are by someone else because we have no means to do that at that stage. Maybe foreign issuer packets should be handled by some separate function.

### CODE REVIEW

File: src/openpgp/keymanager.iced [hunk 1]
```diff
@@ -960,6 +966,35 @@ class KeyManager extends KeyManagerInterface
     @merge_public_omitting_revokes(km2)
     @merge_userids(km2)
 
+  #----------
+
+  find_verified_designated_revoke : (fetcher, cb) ->
+    unless @pgp?
+      return cb()
+
+    sigs = @pgp.primary?._pgp?.get_designated_revocations()
```
Comment (maxtaco) [inline src/openpgp/keymanager.iced]: As discussed in chat, introduce an accessor here.

File: src/openpgp/packet/keymaterial.iced [hunk 2]
```diff
@@ -539,6 +539,34 @@ class KeyMaterial extends Packet
 
     return winner
 
+  #-------------------
+
+  add_designee : (rev_key) ->
+    (@desig_revokers or= []).push rev_key
```
Comment (maxtaco) [inline src/openpgp/packet/keymaterial.iced]: More convenient to store as a map here?

File: src/openpgp/processor.iced [hunk 3]
```diff
@@ -91,8 +91,22 @@ class KeyBlock
...
+
+    if (iid = packet.get_issuer_key_id())?
+      return [iid, bufeq_secure(iid, @primary.get_key_id()), @primary.get_key_id()]
+
+    @warnings.push "Signature is missing an issuer (at packet=#{i})"
+    return [null, false]
+
+  #--------------------
 
+  # TODO: remove this one or the other when we figure out how to
```
Comment (maxtaco) [inline src/openpgp/processor.iced]: what's with this `TODO`? 

Why both `_get_issuer` and `_check_issuer`?

File: src/openpgp/processor.iced [hunk 4]
```diff
@@ -91,8 +91,22 @@ class KeyBlock
...
+  # handle 3rd party signatures.
+  _check_issuer : (i, packet, primary) ->
```
Comment (maxtaco) [inline src/openpgp/processor.iced]: is this being called anymore?

File: src/openpgp/processor.iced [hunk 5]
```diff
@@ -127,9 +119,15 @@ class KeyBlock
...
-      if (issuer_id = @_check_issuer(i, p, @primary))?
+      [issuer_id, matches_ours, expected_ours] = @_get_issuer i, p, @primary
+      if issuer_id?
+        unless matches_ours or p.type is C.sig_types.key_revocation
+          @warnings.push "Skipping signature by another issuer: #{issuer_id?.toString('hex')} != #{expected_ours?.toString('hex')}"
+          continue
+
         p.key = @primary.key
         p.primary = @primary
+        @opts.skip_verify = not matches_ours
```
Comment (maxtaco) [inline src/openpgp/processor.iced]: does this disable all verification for the rest of the processing session?

Comment (maxtaco) [inline src/openpgp/processor.iced]: Can you make a copy of `@opts` and modify it and pass that in? This feels really risky. I don't want to turn verification off by accident.

Comment (zapu) [inline src/openpgp/processor.iced]: Agreed 100%, this is a horrible bug waiting to happen.

File: src/openpgp/packet/signature.iced [hunk 6]
```diff
@@ -257,14 +257,16 @@ class Signature extends Packet
...
 
       else
         err = new Error "cannot verify sigtype #{@type}"
         []
 
     # Now actually check that the signature worked.
-    unless err?
+    # We might want to skip verification of 3rd party signatures for
+    # certain packet types, like key_revocation (pgp designated revoke).
+    unless err? or (@is_third_party and @type is T.key_revocation)
```
Comment (zapu) [inline src/openpgp/packet/signature.iced]: also see here, would this work?

File: src/openpgp/processor.iced [hunk 7]
```diff
@@ -127,9 +119,15 @@ class KeyBlock
...
-      if (issuer_id = @_check_issuer(i, p, @primary))?
+      [issuer_id, matches_ours, expected_ours] = @_get_issuer i, p, @primary
+      if issuer_id?
+        unless matches_ours or p.type is C.sig_types.key_revocation
+          @warnings.push "Skipping signature by another issuer: #{issuer_id?.toString('hex')} != #{expected_ours?.toString('hex')}"
+          continue
+
         p.key = @primary.key
         p.primary = @primary
+        p.is_third_party = not matches_ours
```
Comment (zapu) [inline src/openpgp/processor.iced]: Changed not to mutate `@opts`

File: src/openpgp/packet/signature.iced (hunk 6 above)
Comment (maxtaco) [inline src/openpgp/packet/signature.iced]: yes, it's good. I wonder if we should split this function up in to smaller people. And/or allow early return in the case of errors?  I had previously avoided that, but this function is getting more and more complex

File: src/openpgp/packet/signature.iced [hunk 8]
```diff
@@ -318,8 +320,33 @@ class Signature extends Packet
...
+        when T.direct
+          if fp = @subpacket_index.hashed[S.revocation_key]
+            @primary.add_designee fp
+
+        when T.certificate_revocation
+          if (userid = @data_packets[1].to_userid())?
+            userid.mark_revoked sig
+
+        else
+          err = new Error "Got unknown signature type=#{@type}"
```
Comment (maxtaco) [inline src/openpgp/packet/signature.iced]: have we really covered all cases now?

Comment (maxtaco) [inline src/openpgp/packet/signature.iced]: If not, this might break currently working keys...

Comment (zapu) [inline src/openpgp/packet/signature.iced]: error here ends up as a warning: https://github.com/keybase/kbpgp/blob/master/src/openpgp/processor.iced#L135-L136

but I can remove the `else` block

Comment (maxtaco) [inline src/openpgp/packet/signature.iced]: Ah, ok, then good!

**Pull request merged**